src_c/geometry.c

```diff
@@ -7,6 +7,108 @@
 #endif /* ~__AVX2__ */
 
 #define PYGAMEAPI_GEOMETRY_NUMSLOTS 21
+static PyObject *
+pg_raycast(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    PyObject *startpoint;
+    PyObject *endpoint = NULL;
+    PyObject **farr;
+    PyObject **collisions;
+    Py_ssize_t loop;
+    double angle;
+    double max_dist;
+    double target_pos[2] = {0, 0};
+
+    static char *keywords[] = {"start_pos", "collisions", "endpoint", "angle", "max", NULL};
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "OO|Odd",
+        keywords, &startpoint, &collisions, &endpoint, &angle, &max_dist))
+    {
+        return NULL;
+    }
+
+    farr = PySequence_Fast_ITEMS(collisions);
+
+    if (!PySequence_Check(startpoint)) {
+        return RAISE(PyExc_TypeError, "positions must be sequences");
+    }
+
+    pgLineObject *line = (pgLineObject *)pgLine_Type.tp_new(&pgLine_Type, NULL, NULL);
+    if (line) {
+        line->line.x1 = PyFloat_AsDouble(PySequence_GetItem(startpoint, 0));
+        line->line.y1 = PyFloat_AsDouble(PySequence_GetItem(startpoint, 1));
+    }
+    
+    Py_ssize_t arg0_length = PySequence_Size(startpoint);
+    Py_ssize_t col_length = PySequence_Size(collisions);
+
+    if (arg0_length != 2) {
+        return RAISE(PyExc_TypeError, "incorrect start position size");
+    }
+
+    if (endpoint != NULL) {
+        if (!PySequence_Check(endpoint)) {
+            return RAISE(PyExc_TypeError, "positions must be sequences");
+        }
+        Py_ssize_t arg1_length = PySequence_Size(endpoint);
+        if (arg1_length != 2) {
+            return RAISE(PyExc_TypeError, "incorrect end position size");
+        }
+        line->line.x2 = PyFloat_AsDouble(PySequence_GetItem(endpoint, 0));
+        line->line.y2 = PyFloat_AsDouble(PySequence_GetItem(endpoint, 1));
+    }
+    else {
+        double offset1 = cos(angle * PI / 180) * max_dist;
+        double offset2 = sin(angle * PI / 180) * max_dist;
+        line->line.x2 = line->line.x1 - offset1;
+        line->line.y2 = line->line.y1 - offset2;
```

Comment:
this line->line this will not be needed once you create a pgLineBase struct instead of a PyObject